### ISSUE TITLE
Support Workspace Trust

### ISSUE THREAD

**Issue by lramos15** [VisualSolana/VisualSolana#13]
Hello :wave: I'm from the VS Code team.

Recently, we have been exploring a security feature we refer to as _Workspace Trust_. This feature is intended to centralize and unify a security conscious decision required by a variety of VS Code features. With workspace trust, the user will be able to declare whether or not they trust the folder that is opened in VS Code before these features are executed.

### Why you should care

We want to make sure that those users have a delightful experience with workspace trust and that includes extension authors deciding how much of their extension is supported in an untrusted workspace. Custom editors are special in that if they cannot at least render in an untrusted state then they will be replaced by a lightweight message telling the user to enable trust. This means that users of your extension will not be able to see their editors and there may even be cases of data loss.

![Custom Editors Untrusted](https://user-images.githubusercontent.com/6561887/117475647-29788080-af11-11eb-87e2-56a4f92fd789.gif)

### Workspace Trust experience

You can enable the feature with the following setting `security.workspace.trust.enabled`. Once enabled, you will see the following dialog when opening folders in VS Code.

![Workspace Trust Startup Dialog](https://media.githubusercontent.com/media/microsoft/vscode-docs/vnext/release-notes/images/1_56/trust-dialog.png)

This dialog is important for allowing the user to make a decision early and understand the impact of their decision. Once you understand the feature, you may want to customize when to display the dialog using the setting `security.workspace.trust.startupPrompt`.

You can follow the development of Workspace Trust and provide feedback in [issue #106488](https://github.com/microsoft/vscode/issues/106488).

### Workspace trust API

First off, all of what I am about to say can be found in [issue #120251](https://github.com/microsoft/vscode/issues/120251). That issue will include discussion of the feature and any updates to the feature.

The Workspace Trust extension API is now in stable. This allowed us to release the first cut of our guide for onboarding your extension to Workspace Trust. The API is small, so here is a quick look.

You can declare your extension to provide complete, partial or no support in untrusted workspaces using the `untrustedWorkspaces` capability in `package.json`.

The following example declares that the extension is supported completely in untrusted workspaces. In this case, the extension is enabled in untrusted workspaces.

```json
"capabilities": {
  "untrustedWorkspaces": {
    "supported": true
  }
}
```

The next example declares that the extension is not supported in untrusted workspaces. In this case, the extension is disabled in untrusted workspaces.

```json
"capabilities": {
  "untrustedWorkspaces": {
    "supported": false
  }
}
```

The third option is to declared `limited` support. There are three tools provided to you when you select the `limited` option.

First, if you have a setting that can be configured in the workspace but requires the workspace to be trusted in order to apply the workspace value, then you can include the setting using `restrictedConfigurations` array property in `untrustedWorkspaces` object. Doing so, VS Code will ignore the workspace value of these restricted settings when your extension reads these settings values using the VS Code [Workspace Configuration](https://code.visualstudio.com/api/references/vscode-api#WorkspaceConfiguration) API.

The following example declares the settings that are restricted in untrusted workspaces.

```json
"capabilities": {
  "untrustedWorkspaces": {
    "supported": "limited",
    "restrictedConfigurations": [
      "markdown.styles"
    ]
  }
}
```

Next, you can also check and listen if the current workspace is trusted or not programmatically using the following API:

```ts
export namespace workspace {
  /**
   * When true, the user has explicitly trusted the contents of the workspace.
   */
  export const isTrusted: boolean;
  /**
   * Event that fires when the current workspace has been trusted.
   */
  export const onDidGrantWorkspaceTrust: Event<void>;
}
```

Lastly, you can hide commands or views declaratively with the `isWorkspaceTrusted` context key in your when clauses.

A far more detailed guide on how to onboard which will be updated as we receive feedback can be found in [issue #120251](https://github.com/microsoft/vscode/issues/120251).

### Rollout plan

We are planning on enabling this by default in the near future (most likely next release). To prepare for that day, we want to work with you to allow your editor to work seamlessly alongside the trusted workspace experience

### Our Asks
The main features that should work for custom editors in an untrusted folder is rendering, viewing, and saving (if applicable). After reviewing your codebase, I believe that your extension does certain things that can be dangerous if the folder contents are malicious therefore I believe that `limited` is the best option. When using `limited` you will have to listen to the workspace trust state and disable certain code features (in this case I believe the generation pieces).

Please let me know if you have any question or would like to meet up via voice chat as I would be happy to assist you in getting your extension ready for workspace trust!

**Comment by jgensler8** [comment]
@lramos15 thanks for opening an issues here and taking a look through the codebase!

My understanding of the "trusted workspace" feature (worst case):
1. customer downloads a untrusted repo, they know they don't trust it but want to look around anyway.
2. repo contains a malicious test suite (for example, I have found the Jest test auto-runner very helpful in development of this plugin)
3. customer unknowingly auto-runs the malicious Jest test suite
4. etc... anything is possible here

Trusted workspaces (hopefully) avoid this attack vector by allowing the Jest plugin author to implement a gate on the trusted workspace value. In the example above, the Jest plugin author realizes that it will be easy to implement an attack vector in a test suite so they turn the "auto-run" feature off for untrusted workspaces. However, maybe there is syntax highlighting so part of the extension is useful even in untrusted workspaces. The resulting "support" value would be "limited" because some Jest features are working whereas other are not.

IMHO, the ability to identify these vectors is still on the plugin developer. The main use of the VisualSolana plugin is reading and writing XML. What if there is some ability for buffer overflow when reading malicious XML? I have little knowledge of possible bugs or attack vectors but it seems like a reasonable place to attack.

Is the most "conservative" option to set `"supported": false` to avoid ever running on the plugin on knowingly untrusted code? I would rather do that as this plugin doesn't have many users.

**Comment by lramos15** [comment]
@jgensler8 You are absolutely correct in that is still up to the extension author to understand the possible attack vectors. We will never be able to completely get rid of all of them, but this feature should help mitigate possible risks and make the user more aware that many things execute code in the background when using VS code. You are also correct about how limited works. You disable pieces of the plugin so it can still operate in the untrusted case and re-enable more aggressive features when in the trusted case.

As for whether or not you should adopt `"supported": false`, that is entirely up to you. However, the problem with custom editors and `"supported": false` is that the user won't be able to see your editor at all. This can cause data loss scenarios especially in the initial rollout of workspace trust. Consider this case,
1. You have a dirty file open in your extension
2. You close VS Code 
3. VS Code updates and enables workspace trust
4. You don't trust and close the open tabs because they don't work anyways
5. 🐛 The user has just lost the data that your extension had backed up for them

It's not a common case of course, but the VS Code team wants to ensure that users have a seamless experience hence why I'm reaching out to extension authors to assist with onboarding. If you don't believe there's malicious attack vectors then `supported: true` is fine. My thought reviewing your code was maybe the generation pieces had some sort of execution of user files but I'm not as familiar with your codebase as you are. Let me know if you have any other questions!
